src/rust/engine/src/externs/mod.rs

```diff
@@ -420,3 +421,12 @@ pub enum GeneratorResponse {
   Get(Get),
   GetMulti(Vec<Get>),
 }
+
+pub(crate) fn from_rust_cpython_to_pyo3(value: &PyObject) -> &pyo3::PyAny {
```

Comment:
As @mejrs said, the `&pyo3::PyAny` is only valid while the GIL is held: the simplest option would be to return an owned `PyAny` (or whatever the pyo3 equivalent is) by cloning it.

More complicated would be to continue to return a reference, but require that the GIL is already held before calling the function by accepting the `py` token as an argument: then the signature might look something like:
```
pub(crate) fn from_rust_cpython_to_pyo3<'p, 'o>(py: Python<'p>, value: &'o PyObject) -> &'p pyo3::PyAny {
  ..
}
```

Which of those is preferable depends on whether the GIL is already likely to be held at callsites? But I'm also not familiar enough with `pyo3`'s API to know.

Reply:
The owned `PyAny` is currently bundled up in a smart pointer `Py<PyAny>`, which also doesn't have a lifetime tied to the GIL.

Reply:
That said, I would probably go with `&PyAny` and require the GIL to be held at callsites. `Python::with_gil` had some performance overhead so it seems undesirable to do it unconditionally inside a low level conversion function like this. If you took the GIL token as an argument, you could do a bulk conversion of many objects without having to acquire and release the GIL for each one.